### PR TITLE
Update specification

### PR DESCRIPTION
**Pull request by mppf**
This PR addresses the easier spec updates I need to do but leaves as
future work the nilable types section and error handling sections.
However it adds these sections so we can start referring to them.

* Adds Error Handling placeholder section
* Adds Nilable Classes placeholder section
* Updates for undecorated-class-is-generic, describe error for multiple
  memory management strategies, casting to replace it
* Updates for default intent for owned/shared being `const ref`
* Describes enum/record/class as type
* Describes prototype modules

Reviewed by @vasslitvinov - thanks!

### CODE REVIEW

File: spec/Generics.tex [hunk 1]
```diff
@@ -475,12 +467,84 @@ \section{Generic Types}
...
+\item
+\chpl{class} can instantiate with any non-nilable class using any
+memory management strategy
+
+\item
+\chpl{class?} can instantiate with any class using any memory managemeent
+strategy but will use the nilable variant of that class in an
+instantiation. When used as an argument type, a value of non-nilable class
+type will be implicitly converted to the nilable type on the call. As a
+result, in practice, this type can accept any class type.
```
Comment (vasslitvinov) [inline spec/Generics.tex]: `As a result, a formal of the type \chpl{class?}  can accept an actual of any class type.`

File: spec/Generics.tex [hunk 2]
```diff
@@ -475,12 +467,84 @@ \section{Generic Types}
...
+\index{enum (generic type)@\chpl{enum} (generic type)}
+\index{class (generic type)@\chpl{class} (generic type)}
+\index{record (generic type)@\chpl{record} (generic type)}
+
+The types \chpl{integral}, \chpl{numeric} and \chpl{enum}
+are generic types that can only be instantiated with, respectively, the
+signed and unsigned integral types, all of the numeric types, and
+all enumerated types. The type \chpl{enumerated} is currently available
+as a synonym for \chpl{enum}.
+
```
Comment (vasslitvinov) [inline spec/Generics.tex]: I suggest this reorg of the following 3 paragraphs i.e. until \begin{itemize} :

```
The type \chpl{record} can be instantiated with any record type.

The memory management strategies \chpl{owned}, \chpl{shared},
\chpl{borrowed}, and \chpl{unmanaged} (see~\rsec{Class_Types})
are also generic types that can be instantiated with any class
using that memory management strategy.

The types \chpl{class} and \chpl{class?}, on their own or
in combination with memory management strategies, are also
generic types. They can be instantiated as follows:
```

File: spec/Modules.tex [hunk 3]
```diff
@@ -42,6 +45,26 @@ \section{Module Definitions}
...
+\begin{itemize}
+
+  \item errors that are not handled will terminate the program
+        (see~\rsec{Error_Handling})
+
+  \item methods on nilable class instances can be called - these will
+        implicitly convert to non-nilable class instances
+        (see~\rsec{Methods_On_Nilable_In_Prototype_Modules})
+
+\end{itemize}
```
Comment (vasslitvinov) [inline spec/Modules.tex]: +1

File: spec/Modules.tex [hunk 4]
```diff
@@ -333,7 +357,8 @@ \subsection{Module Initialization}
 
 Module initialization occurs at program start-up.  All top-level
 statements in a module other than function and type declarations are
-executed during module initialization.
+executed during module initialization. Top-level modules and submodules
+that are not referred to will not be initialized.
```
Comment (vasslitvinov) [inline spec/Modules.tex]: "that are not referred to" has ambiguous applicability. How about:

`Modules that are not referred to, including both top-level modules and submodules, will not be initialized.`

File: spec/Types.tex [hunk 5]
```diff
@@ -66,6 +68,13 @@ \section{Primitive Types}
   `imag' primitive-type-parameter-part[OPT]
   `complex' primitive-type-parameter-part[OPT]
   `string'
+  `enum'
```
Comment (vasslitvinov) [inline spec/Types.tex]: `enumerated` ?

File: spec/Types.tex [hunk 6]
```diff
@@ -385,9 +394,21 @@ \subsection{Class Types}
...
+  \item \chpl{shared C} and \chpl{shared C?}
+  \item \chpl{borrowed C} and \chpl{borrowed C?}
+  \item \chpl{unmanaged C} and \chpl{unmanaged C?}
+\end{itemize}
+
+The variants with a question mark, such as \chpl{owned C?}, can store
+\chpl{nil} (see~\rsec{Nilable_Classes}). The keywords \chpl{owned},
+\chpl{shared}, \chpl{borrowed}, and \chpl{unmanaged} indicate the memory
+management startegy used for the class. When none is specified, as with
+\chpl{C}, the class is considered to have generic memory management
```
Comment (vasslitvinov) [inline spec/Types.tex]: let's add \chpl{C?}: `, as with \chpl{C} and \chpl{C?}, `

File: spec/Types.tex [hunk 7]
```diff
@@ -385,9 +394,21 @@ \subsection{Class Types}
...
+\begin{itemize}
+  \item \chpl{C} and \chpl{C?}
+  \item \chpl{owned C} and \chpl{owned C?}
+  \item \chpl{shared C} and \chpl{shared C?}
+  \item \chpl{borrowed C} and \chpl{borrowed C?}
+  \item \chpl{unmanaged C} and \chpl{unmanaged C?}
+\end{itemize}
+
+The variants with a question mark, such as \chpl{owned C?}, can store
+\chpl{nil} (see~\rsec{Nilable_Classes}). The keywords \chpl{owned},
```
Comment (vasslitvinov) [inline spec/Types.tex]: `, whereas the variants without it, such as \chpl{owned C}, can't`

**Comment by mppf** [comment]
> After you merge, I will follow up to explain that casts to class et al. can apply to types.

Thanks!

> I thought to also include the combinations of mem management + class + ? in the grammar for primitive-type. However, to avoid doing, so we can consider them "not primitive".

I thought about that too, but `integral` and `enumerated` are documented elsewhere but not listed there also. It seems it'd take some effort to write the grammar rules for management + class + ? though. Anyway I think we should do that in a follow-up PR, if at all.

**Comment by vasslitvinov** [comment]
P.S. How about remove `prototype` from "The following identifiers are keywords reserved for future use", Lexical_Structure.tex :244.

File: spec/Types.tex (hunk 5 above)
Comment (mppf) [reply spec/Types.tex]: It wasn't listed here before, and I don't want to add it now :)

Comment (mppf) [reply spec/Types.tex]: (That's mostly because I'm expecting we will deprecate it in favor of just using `enum`).

**Pull request merged**